### ISSUE TITLE
Histogram Metric for multi instance latency

### ISSUE THREAD

**Issue by AshwinAmbal** [triton-inference-server/server#7672]
**Is your feature request related to a problem? Please describe.**
This issue is similar to the one mentioned here: https://github.com/triton-inference-server/server/issues/7287. I'd like to file an issue for histogram metric in Triton core. I remember this being mentioned as being in the backlog from the previous issue but would like to have this for tracking purposes. 

Currently it isn't possible to calculate 95th or 99th percentile latencies when we deploy multiple Triton servers to host models. This also affects scaling decisions from being imprecise. 

**Describe the solution you'd like**
Use histogram-quantile instead of summaries as the Prometheus metric exporter: https://prometheus.io/docs/practices/histograms/#:~:text=The%20essential%20difference%20between%20summaries,the%20server%20side%20using%20the

**Describe alternatives you've considered**
I've tried the wrong way of doing it (Avg of P95 across different Triton servers) but that is not the true p95. 

I've also tried using the Avg queue time/latency (keeping those very low) instead but that doesn't help with dealing with tail latencies during spikes.

**Comment by rmccorm4** [comment]
Hi @AshwinAmbal, thanks for adding a ticket for tracking!

CC @yinggeh @harryskim @statiraju for viz

**Comment by AshwinAmbal** [comment]
Hi @rmccorm4 @yinggeh @harryskim @statiraju 
I had a question on the Histogram support added in response to this issue. 
Does the histogram metric shown here: https://github.com/triton-inference-server/server/blob/main/docs/user_guide/metrics.md#histograms, apply to end to end latencies that we get in Traditional ML systems as well?

If we have a normal 1 request has 1 response type ML system, will the metric `nv_inference_first_response_histogram_ms` give the end-to-end latency which can be used to aggregate across multiple instances?

TLDR; for my specific case, can I consider that the above metric is the "histogram" version of the summary metric here: `nv_inference_request_summary_us`?

Thanks

**Comment by yinggeh** [comment]
Hi @AshwinAmbal. I am happy to answer your questions.

Currently, TTFT histogram metric will only appear in decoupled models. If your traditional ML model is decoupled you should see this metric. If your model is not decoupled, are you able to modify it into a decoupled model but always returns one response?

See https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/decoupled_models.html

**Comment by AshwinAmbal** [comment]
@yinggeh thanks for the speedy response. 

I may not be able to use decoupled mode easily because it requires [gRPC Bidirectional streaming](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/decoupled_models.html#running-inference-on-decoupled-models) from the client side which would be a limitation for different clients to connect to my service. 

Btw, my goal is simple: I want to interpret P50, P95 and P99 latencies across all my Triton servers running in K8s together. 
Currently, each server emits the summary latency `nv_inference_request_summary_us` which is not aggregatable across multiple Triton servers [as it doesn't emit the distribution itself and only calculates the percentiles locally considering only the requests received by the server itself]. 

Similar work had been done on the Triton Python backend as seen in #7287 and hence I was requesting the same to be added to Triton core as well so that we can use it in general. 

Any idea if this is still being planned to be delivered or if there is another way to achieve the same?

**Comment by yinggeh** [comment]
@AshwinAmbal Correct. You cannot directly aggregate percentiles from a Summary metric. To compute percentiles using a Histogram, refer to [Prometheus’ histogram_quantile() function](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile).

When I added `nv_inference_first_response_histogram_ms` metric to Triton, two new metrics—TTFT and TPOT (not yet availble)—were intended to introduced for LLM models. I expanded the coverage to decoupled models and renamed "time to first token" to "time to first response". While this metric was designed for decoupled models, it should also work with non-decoupled models, though the naming might be odd.

If this is urgent, you can try the following modification and rebuild the server. Let me know what you think.
```
--- a/src/metric_model_reporter.cc
+++ b/src/metric_model_reporter.cc
@@ -308,10 +308,8 @@ MetricModelReporter::InitializeHistograms(
   // Update MetricReporterConfig::metric_map_ for new histograms.
   // Only create response metrics if decoupled model to reduce metric output
   if (config_.latency_histograms_enabled_) {
-    if (config_.is_decoupled_) {
       histogram_families_[kFirstResponseHistogram] =
           &Metrics::FamilyFirstResponseDuration();
-    }
   }
 
   for (auto& iter : histogram_families_) {
```

**Comment by yinggeh** [comment]
> Any idea if this is still being planned to be delivered or if there is another way to achieve the same?

Do you have a list of latency Histograms for your deployment?

**Comment by AshwinAmbal** [comment]
@yinggeh This is important for auto-scaling Triton servers in Kubernetes while monitoring latency as the metric. Currently, this isn't possible. 

Ideally what I would like to see is all the [summary metrics](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/metrics.html#summaries) supported today also available as histogram type.  

In particular, I am interested mainly in `nv_inference_request_summary_us` as a histogram metric.

I can try the workaround that you suggested to see its work but it would be nice to have an actual solution for this as it seems like a foundational metric to me. WDYT?

**Comment by AshwinAmbal** [comment]
I believe there is only one histogram metric available today: `nv_inference_first_response_histogram_ms` according to Triton docs [here](https://github.com/triton-inference-server/server/blob/main/docs/user_guide/metrics.md#histograms) and it only works for LLM/Decoupled models from the description from @yinggeh 

About their working, I'll have to redirect you to a Triton maintainer. I haven't begun using the histogram metric yet because the one I want isn't available yet. 

Gentle bump to @yinggeh for this and my previous query please. 

Thanks

**Comment by geraldstanje** [comment]
@AshwinAmbal i use ```--metrics-config histogram_latencies=true``` and dont see any histogram. do you see nv_inference_first_response_histogram_ms?

**Comment by yinggeh** [comment]
Hi @geraldstanje. Please find details on decoupled models here https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/decoupled_models.html. In short, decoupled models can send zero or multiple responses for a request.

To enable decoupled mode, you need to config model properly with
```
model_transaction_policy {
  decoupled: True
}
```

**Comment by amannedd** [comment]
@yinggeh could you explain a bit more about the reasoning for only enabling histogram metrics for decoupled models? It would be useful to have in order to track P99 e2e latency of our Triton server given that the available metrics only allow us to see the average latency. Thanks.

**Comment by yinggeh** [comment]
Hi @amannedd. Sorry for the late response. By that time, we received requests to add two histogram metrics for LLM models (TTFT and TPOT). LLM models in nature are decoupled models in Triton.

If you feel like some particular histogram metrics for non-decoupled models are useful in your production, I encourage to post a feature request in https://github.com/triton-inference-server/server/issues with expected behavior and output. Thanks.

**Comment by AshwinAmbal** [comment]
@yinggeh any update on the plan to add histogram support in Triton for traditional ML models rather than for LLM's? TTFT and TPOT are not relevant to traditional ML models and aren't useful unfortunately.

**Comment by geraldstanje** [comment]
i have the same question - all type of models should have histogram support

**Comment by zbennett10** [comment]
I've implemented histogram equivalents for all 5 summary latency metrics (request, queue, compute_input, compute_infer, compute_output). Changes span core + server repos. PRs incoming! :D

**Comment by ziollek** [comment]
@yinggeh - is there any chance to release change prepared by @zbennett10 ?

**Comment by yinggeh** [comment]
@Vinya567 @mudit-eng for visibility
